apps/system/js/places.js

```diff
@@ -92,12 +92,76 @@
         window.addEventListener('apploaded', this);
 
         asyncStorage.getItem('top-sites', results => {
-          this.topSites = results || [];
+          this.topSites = this._removeDupes(results || []);
           resolve();
         });
       });
     },
 
+    /**
+     * Remove duplicated entries.
+     * @param {Array} topSites array of places object from asyncStorage()
+     * @return {Array} of places object without duplicated entries
+     */
+    _removeDupes: function(ts) {
+      if (!ts) {
+          return [];
+      }
+
+      // First we count the URLs and we update a counter
+      var counter = {};
+      ts.forEach(e => {
+        if (!('url' in e)) {
+          return;
+        }
+
+        if (e.url in counter) {
+          counter[e.url] += 1;
+        } else {
+          counter[e.url]  = 1;
+        }
+      });
+
+      // Duplicated entries are URLs with counter > 1
+      var dupes = [];
+      for (var id in counter) {
+        if (counter[id] > 1) {
+          dupes.push(id);
+        }
+      }
+
+      // checkTopSites() does the orderding by decreasing 'frecency'
+      // So for each dupe, we remove all the entries except the first one
+      dupes.forEach(dupe => {
+        var nbTS = ts.length;
+        var dupeId = 0;
+        for (var i = 0; i < nbTS; i++) {
+          if (!ts[i]) {
+            continue;
+          }
+
+          if (ts[i].url === dupe) {
+            dupeId++;
+            if (dupeId > 1) {
+              console.log('topSites already contains', dupe, 'at index', i,
+                          'removing ...');
+              delete ts[i];
+            }
+          }
+        }
+      });
+
+      // Copy back into a new array to avoid null in the middle
+      var rv = [];
+      ts.forEach(t => {
+        if (t) {
+          rv.push(t);
+        }
+      });
+
+      return rv;
+    },
+
```

Comment:
So this is a fairly long and hard to understand function, I think it could be a little simpler, how about copying the array and keeping track of copied urls, so

```
var copy = [];
var copied = {};
ts.forEach(function(place) { 
  if (place.url && !(place.url in copied)) { 
    copied[place.url] = true;
    copy.push(place);
  }
});
```
Would that work?